torch/csrc/autograd/python_function.cpp

```diff
@@ -1055,3 +1069,67 @@ std::shared_ptr<PyFunction> THPFunction_asFunction(THPFunction* self)
   Py_INCREF((PyObject*)self);
   return std::shared_ptr<PyFunction>(&self->cdata, Decref());
 }
+
+namespace torch { namespace autograd {
+
+using memo_map = std::unordered_map<Node*, variable_list>;
+
+variable_list interpret_node(std::shared_ptr<Node>, input_map&, memo_map&);
+
+std::shared_ptr<Variable> interpret_output(Output output, input_map& inputs, memo_map& memo) {
+    auto vs = interpret_node(output.node, inputs, memo);
+    return vs.at(output.output_nr);
+}
+
+variable_list interpret_node(std::shared_ptr<Node> node, input_map& inputs, memo_map& memo) {
+    variable_list input_vars;
+    auto num_args = node->inputs.size();
+    input_vars.reserve(num_args);
+    for (auto input : node->inputs) {
+        input_vars.emplace_back(interpret_output(input, inputs, memo));
+    }
+    if (auto n = dynamic_cast<InputNode*>(node.get())) {
+        return {inputs.at(n)};
+    } else if (auto n = dynamic_cast<PyNode*>(node.get())) {
+        auto& cls = n->pyobj;
+        // Massage variables into form where we can THPFunction_apply it.
+        // While in principle we can avoid putting things into Python
+        // and then taking them out again, doing so seems to require an excess
+        // of faffing about to optimize a codepath that is already going to
+        // fundamentally be slow (since it calls into Python.) NOT. WORTH. IT.
+
+        PyObject* input_objs = PyTuple_New(num_args);
+        for (size_t i = 0; i < num_args; i++) {
+          PyTuple_SET_ITEM(input_objs, i, THPVariable_Wrap(input_vars.at(i)));
+        }
+        THPObjectPtr output_objs{ THPFunction_apply(cls, input_objs) };
+        _ensure_tuple(output_objs);
+        auto num_outputs = PyTuple_GET_SIZE(output_objs.get());
+        variable_list output_vars;
+        output_vars.reserve(num_outputs);
+        for (Py_ssize_t i = 0; i < num_outputs; i++) {
+          PyObject *output_obj = PyTuple_GET_ITEM(output_objs.get(), i);
+          if (!THPVariable_Check(output_obj)) {
+            throw std::logic_error("Unwrapped variable from THPFunction_apply");
```

Comment:
I can't see how this error really tells what happened. Perhaps "THPFunction_apply returned a non-Variable object"?